src/spatialdata/_core/operations/aggregate.py

```diff
@@ -0,0 +1,261 @@
+    target_coordinate_system: str = "global",
+    **kwargs: Any,
+) -> ad.AnnData:
+    """
+    Aggregate values by given region.
+
+    Parameters
+    ----------
+    values
+        Values to aggregate.
+    by
+        Regions to aggregate by.
+    id_key
+        Key to group observations in `values` by. E.g. this could be transcript id for points.
+        Defaults to `FEATURE_KEY` for points, required for shapes.
+    value_key
+        Key to aggregate values by. This is the key in the values object.
+        If nothing is passed here, assumed to be a column of ones.
+        For points, this could be probe intensity.
+    agg_func
+        Aggregation function to apply over point values, e.g. "mean", "sum", "count".
+        Passed to :func:`pandas.DataFrame.groupby.agg` or from :func:`xrspatial.zonal_stats`
+        according to the type of `values`.
+    target_coordinate_system
+        Coordinate system to transform to before aggregating.
+    kwargs
+        Additional keyword arguments to pass to :func:`xrspatial.zonal_stats`.
+
+    Returns
+    -------
+    AnnData of shape (by.shape[0], values[id_key].nunique())])
+    """
+    # get schema
+    by_type = get_model(by)
+    values_type = get_model(values)
+
+    # get transformation between coordinate systems
+    by_transform: BaseTransformation = get_transformation(by, target_coordinate_system)  # type: ignore[assignment]
+    values_transform: BaseTransformation = get_transformation(
+        values,
+        target_coordinate_system,  # type: ignore[assignment]
+    )
+    if not ((by_transform == values_transform) and (values_transform is Identity)):
+        by = transform(by, by_transform)
+        values = transform(values, values_transform)
+
+    # dispatch
+    if by_type is ShapesModel:
+        if values_type is PointsModel:
+            return _aggregate_points_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        if values_type is ShapesModel:
+            return _aggregate_shapes_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+    if by_type is Labels2DModel and values_type is Image2DModel:
+        return _aggregate_image_by_labels(values, by, agg_func, **kwargs)
+    raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+
+
+def _aggregate_points_by_shapes(
+    points: ddf.DataFrame | pd.DataFrame,
+    shapes: gpd.GeoDataFrame,
+    id_key: str | None = None,
+    *,
+    value_key: str | None = None,
+    agg_func: str | list[str] = "count",
+) -> ad.AnnData:
+    # Have to get dims on dask dataframe, can't get from pandas
+    dims = get_axes_names(points)
+    # Default value for id_key
+    if id_key is None:
+        id_key = points.attrs[PointsModel.ATTRS_KEY][PointsModel.FEATURE_KEY]
+
+    if isinstance(points, ddf.DataFrame):
+        points = points.compute()
```

Comment:
I think we should add a note to the docstring to say this will bring lazy points into memory. That could be an unpleasant surprise for a user working with large, lazy data.